cmd/utils.go

```diff
@@ -157,3 +162,70 @@ func getLookupType(l string) minio.BucketLookupType {
 	}
 	return minio.BucketLookupAuto
 }
+
+// sumMD5Base64 calculate md5sum for an input byte array, returns base64 encoded.
+func sumMD5Base64(data []byte) string {
+	hash := md5.New()
+	hash.Write(data)
+	return base64.StdEncoding.EncodeToString(hash.Sum(nil))
+}
+
+// struct representing object prefix and sse keys association.
+type prefixSSEPair struct {
+	prefix string
+	sseKey string
+}
+
+// parse list of space separated alias/prefix=sse key values entered on command line and
+// construct a map of alias to prefix and sse pairs
+func parseEncryptionKeys(ssekeys string) (encMap map[string][]prefixSSEPair, err *probe.Error) {
+	encMap = make(map[string][]prefixSSEPair)
+	if ssekeys == "" {
+		return
+	}
+	fields := strings.Split(ssekeys, ",")
+	r := regexp.MustCompile("([^=\\s]+)=([^\\s]+)")
+	for _, field := range fields {
+		field := strings.TrimSpace(field)
+		matches := r.FindStringSubmatch(field)
+		if len(matches) != 3 {
+			return nil, probe.NewError(errors.New("sse-c prefix should be of the form prefix=key"))
+		}
+		key := matches[1]
+		value := matches[2]
+		alias, _ := url2Alias(key)
+		if len(value) != 32 {
+			return nil, probe.NewError(errors.New("sse-c key should be 32 bytes long"))
+		}
+		prefix := strings.TrimSpace(key)
+		if _, ok := encMap[alias]; !ok {
+			encMap[alias] = make([]prefixSSEPair, 0)
+		}
+		ps := prefixSSEPair{prefix: prefix, sseKey: value}
+		encMap[alias] = append(encMap[alias], ps)
+	}
+	// sort encryption keys in descending order of prefix length
+	for _, encKeys := range encMap {
+		sort.Sort(byPrefixLength(encKeys))
```

Comment:
you don't want to save them back into encMap after sorting?

Reply:
In place sorting - tested to double check.
```
➜  mc git:(encrypt) ✗ mc cat myminio-ssl/testbucket/obj --insecure --encrypt-key "myminio/testbucket=32byteslongsecretkeymustgenerate, myminio/testbucket/a/b=32byteslongsecretkeymustgenerat1,myminio/testbucket/a/b/c=32byteslongsecretkeymustgenerat2,myminio/testbucket/a/b/c/d=32byteslongsecretkeymustgenerat4"
encryption keys after parse map[myminio:[{myminio/testbucket/a/b/c/d 32byteslongsecretkeymustgenerat4} {myminio/testbucket/a/b/c 32byteslongsecretkeymustgenerat2} {myminio/testbucket/a/b 32byteslongsecretkeymustgenerat1} {myminio/testbucket 32byteslongsecretkeymustgenerate}]]
```